flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -530,14 +541,50 @@ private void registerSlot(
 			removeSlot(slotId);
 		}
 
-		TaskManagerSlot slot = new TaskManagerSlot(
+		final TaskManagerSlot slot = createAndRegisterTaskManagerSlot(slotId, resourceProfile, taskManagerConnection);
+
+		final PendingTaskManagerSlot pendingTaskManagerSlot;
+
+		if (allocationId == null) {
+			pendingTaskManagerSlot = findExactlyMatchingPendingTaskManagerSlot(resourceProfile);
```

Comment:
Why use exactly matching here but isMatching in other place? I think we should make the methods for finding pending task manager slots extendable so user can define their own matching algorithm according to need.

Reply:
Because we use the `ResourceProfile` at the moment as a key for fulfilling pending task manager slots. Therefore, we need to find an exactly matching resource profile. In the future we could introduce an ID to identify newly started slots.

Think for example that you have two pending slots with resource profile `(1,1)` and `(2,2)`. Now a slot with `(2,2)` is registered. You don't want to complete the pending task manager slot `(1,1)` (even though it would fulfill the requirements), because the slot `(1,1)`, won't be able to complete the remaining pending task manager slot.

Making the matching algorithm pluggable should be a separate issue.